File: ios/Classes/HTKLineContainerView.swift
//
//  HTKLineContainerView.swift
//  Base64
//
//  Created by hublot on 2020/8/26.
//

import UIKit

class HTKLineContainerView: UIView {
    
    var configManager = HTKLineConfigManager()
    
    @objc var onDrawItemDidTouch: RCTBubblingEventBlock?

    // Fired when the user taps the hover price pill on the right side (long-press selector).
    // JS receives: { price: number }
    @objc var onNewOrder: RCTBubblingEventBlock?
    
    @objc var onDrawItemComplete: RCTBubblingEventBlock?

    // Continuously fired while a drawing (line/text/etc.) is being moved.
    // JS receives the current index, drawType, pointList, and text (if any).
    @objc var onDrawItemMove: RCTBubblingEventBlock?
    
    @objc var onDrawPointComplete: RCTBubblingEventBlock?
    
    // Called when user scrolls to the left edge (request older candles)
    @objc var onEndReached: RCTBubblingEventBlock?
    
    @objc var optionList: String? {
        didSet {
            guard let optionList = optionList else {
                return
            }
            
            RNKLineView.queue.async { [weak self] in
                do {
                    guard let optionListData = optionList.data(using: .utf8),
                          let optionListDict = try JSONSerialization.jsonObject(with: optionListData, options: .allowFragments) as? [String: Any] else {
                        return
                    }
                    self?.configManager.reloadOptionList(optionListDict)
                    DispatchQueue.main.async {
                        guard let self = self else { return }
                        self.reloadConfigManager(self.configManager)
                    }
                } catch {
                    print("Error parsing optionList: \(error)")
                }
            }
        }
    }

    // Lightweight data-only update: replace modelArray without reloading full optionList.
    // Accepts the same modelArray JSON you normally embed inside optionList.
    @objc var modelArray: String? {
        didSet {
            guard let modelArray = modelArray else {
                return
            }
            
            RNKLineView.queue.async { [weak self] in
                do {
                    guard let data = modelArray.data(using: .utf8),
                          let list = try JSONSerialization.jsonObject(with: data, options: .allowFragments) as? [[String: Any]] else {
                        return
                    }
                    self?.configManager.modelArray = HTKLineModel.packModelArray(list)
                    DispatchQueue.main.async {
                        guard let self = self else { return }
                        self.klineView.reloadContentSize()
                        self.klineView.scrollViewDidScroll(self.klineView)
                    }
                } catch {
                    print("Error parsing modelArray: \(error)")
                }
            }
        }
    }

    lazy var klineView: HTKLineView = {
        let klineView = HTKLineView.init(CGRect.zero, configManager)
        klineView.containerView = self
        return klineView
    }()
    
    lazy var shotView: HTShotView = {
        let shotView = HTShotView.init(frame: CGRect.zero)
        shotView.dimension = 100
        return shotView
    }()

    func setupChildViews() {
        klineView.frame = bounds
        let superShotView = reactSuperview()?.reactSuperview()?.reactSuperview()
        superShotView?.reactSuperview()?.addSubview(shotView)
        shotView.shotView = superShotView
        shotView.reactSetFrame(CGRect.init(x: 50, y: 50, width: shotView.dimension, height: shotView.dimension))
    }

    override var frame: CGRect {
        didSet {
	        setupChildViews()
        }
    }
    
    override func reactSetFrame(_ frame: CGRect) {
        super.reactSetFrame(frame)
        setupChildViews()
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        addSubview(klineView)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func reloadConfigManager(_ configManager: HTKLineConfigManager) {
        
        configManager.onDrawItemDidTouch = { [weak self] (drawItem, drawItemIndex) in
            self?.configManager.shouldReloadDrawItemIndex = drawItemIndex
            guard let drawItem = drawItem, let colorList = drawItem.drawColor.cgColor.components else {
                self?.onDrawItemDidTouch?([
                    "index": drawItemIndex,
                    "shouldReloadDrawItemIndex": drawItemIndex,
                    "id": drawItem?.uid ?? ""
                ])
                return
            }
            self?.onDrawItemDidTouch?([
                "index": drawItemIndex,
                "shouldReloadDrawItemIndex": drawItemIndex,
                "id": drawItem.uid,
                "drawColor": colorList,
                "drawLineHeight": drawItem.drawLineHeight,
                "drawDashWidth": drawItem.drawDashWidth,
                "drawDashSpace": drawItem.drawDashSpace,
                "drawIsLock": drawItem.drawIsLock
            ])
        }
        configManager.onDrawItemComplete = { [weak self] (drawItem, drawItemIndex) in
            guard let this = self, let drawItem = drawItem else {
                self?.onDrawItemComplete?([AnyHashable: Any].init())
                return
            }

            func colorToInt(_ color: UIColor) -> Int {
                var r: CGFloat = 0
                var g: CGFloat = 0
                var b: CGFloat = 0
                var a: CGFloat = 0
                color.getRed(&r, green: &g, blue: &b, alpha: &a)
                let ai = Int(a * 255.0) << 24
                let ri = Int(r * 255.0) << 16
                let gi = Int(g * 255.0) << 8
                let bi = Int(b * 255.0)
                return ai | ri | gi | bi
            }

            var pointArray = [[String: Any]]()
            for point in drawItem.pointList {
                pointArray.append([
                    "x": point.x,
                    "y": point.y
                ])
            }
            
            this.onDrawItemComplete?([
                "index": drawItemIndex,
                "id": drawItem.uid,
                "drawType": drawItem.drawType.rawValue,
                "drawColor": colorToInt(drawItem.drawColor),
                "drawLineHeight": drawItem.drawLineHeight,
                "drawDashWidth": drawItem.drawDashWidth,
                "drawDashSpace": drawItem.drawDashSpace,
                "drawIsLock": drawItem.drawIsLock,
                "pointList": pointArray,
                "text": drawItem.text,
                "textColor": colorToInt(drawItem.textColor),
                "textBackgroundColor": colorToInt(drawItem.textBackgroundColor),
                "textCornerRadius": drawItem.textCornerRadius,
                // Expose per-item text font size (falls back to candleTextFontSize when 0 on native).
                "fontSize": drawItem.textFontSize > 0 ? drawItem.textFontSize : configManager.candleTextFontSize,
                // Expose candleMarker position ("top" | "bottom") so JS can persist it.
                "position": drawItem.position
            ])
        }
        configManager.onDrawItemMove = { [weak self] (drawItem, drawItemIndex) in
            guard let drawItem = drawItem else {
                return
            }

            var pointArray = [[String: Any]]()
            for point in drawItem.pointList {
                pointArray.append([
                    "x": point.x,
                    "y": point.y
                ])
            }
            
            self?.onDrawItemMove?([
                "index": drawItemIndex,
                "id": drawItem.uid,
                "drawType": drawItem.drawType.rawValue,
                "pointList": pointArray,
                "text": drawItem.text,
                "position": drawItem.position
            ])
        }
        configManager.onDrawPointComplete = { [weak self] (drawItem, drawItemIndex) in
            guard let drawItem = drawItem else {
                return
            }
            self?.onDrawPointComplete?([
                "pointCount": drawItem.pointList.count
            ])
        }
        
        let reloadIndex = configManager.shouldReloadDrawItemIndex
        if reloadIndex >= 0, reloadIndex < klineView.drawContext.drawItemList.count {
            let drawItem = klineView.drawContext.drawItemList[reloadIndex]
            drawItem.drawColor = configManager.drawColor
            drawItem.drawLineHeight = configManager.drawLineHeight
            drawItem.drawDashWidth = configManager.drawDashWidth
            drawItem.drawDashSpace = configManager.drawDashSpace
            drawItem.drawIsLock = configManager.drawIsLock
            if (configManager.drawShouldTrash) {
                configManager.shouldReloadDrawItemIndex = HTDrawState.showPencil.rawValue
                klineView.drawContext.drawItemList.remove(at: reloadIndex)
                configManager.drawShouldTrash = false
            }
            klineView.drawContext.setNeedsDisplay()
        }

        // If a serialized drawing list was provided from React Native,
        // rebuild the native drawItemList from it (pre-insert drawings).
        if let rawList = configManager.drawItemList {
            klineView.drawContext.clearDrawItemList()
            for item in rawList {
                guard let points = item["pointList"] as? [[String: Any]],
                      points.count > 0,
                      let firstPoint = points.first
                else {
                    continue
                }

                let rawType = (item["drawType"] as? Int) ?? 0
                guard let drawType = HTDrawType(rawValue: rawType) else {
                    continue
                }
                
                guard let x = firstPoint["x"] as? CGFloat else {
                    continue
                }
                
                // Optional position for candleMarker ("top" | "bottom").
                let position = (item["position"] as? String) ?? "bottom"
                
                // For candleMarker, derive Y from the candle body at this timestamp so
                // JS can omit `y` in pointList and only provide the horizontal anchor.
                let startY: CGFloat
                if drawType == .candleMarker {
                    let isTop = position.lowercased() == "top"
                    startY = candleMarkerBodyValue(forX: x, useTop: isTop)
                } else {
                    guard let y = firstPoint["y"] as? CGFloat else {
                        continue
                    }
                    startY = y
                }
                
                let uid = item["id"] as? String
                let drawItem = HTDrawItem(drawType, CGPoint(x: x, y: startY), uid: uid)
                drawItem.position = position

                // Remaining points
                if points.count > 1 {
                    for pointDict in points.dropFirst() {
                        guard let px = pointDict["x"] as? CGFloat,
                              let py = pointDict["y"] as? CGFloat
                        else {
                            continue
                        }
                        drawItem.pointList.append(CGPoint(x: px, y: py))
                    }
                }

                // Style properties, with fallback to global draw config
                if let colorInt = item["drawColor"] as? Int,
                   let uiColor = RCTConvert.uiColor(colorInt) {
                    drawItem.drawColor = uiColor
                } else {
                    drawItem.drawColor = configManager.drawColor
                }
                if let lineHeight = item["drawLineHeight"] as? CGFloat {
                    drawItem.drawLineHeight = lineHeight
                } else {
                    drawItem.drawLineHeight = configManager.drawLineHeight
                }
                if let dashWidth = item["drawDashWidth"] as? CGFloat {
                    drawItem.drawDashWidth = dashWidth
                } else {
                    drawItem.drawDashWidth = configManager.drawDashWidth
                }
                if let dashSpace = item["drawDashSpace"] as? CGFloat {
                    drawItem.drawDashSpace = dashSpace
                } else {
                    drawItem.drawDashSpace = configManager.drawDashSpace
                }
                if let isLock = item["drawIsLock"] as? Bool {
                    drawItem.drawIsLock = isLock
                } else {
                    drawItem.drawIsLock = configManager.drawIsLock
                }

                if let text = item["text"] as? String {
                    drawItem.text = text
                }

                if let textColorInt = item["textColor"] as? Int,
                   let textColor = RCTConvert.uiColor(textColorInt) {
                    drawItem.textColor = textColor
                } else {
                    drawItem.textColor = configManager.drawTextColor
                }
                if let textBackgroundColorInt = item["textBackgroundColor"] as? Int,
                   let textBackgroundColor = RCTConvert.uiColor(textBackgroundColorInt) {
                    drawItem.textBackgroundColor = textBackgroundColor
                } else {
                    drawItem.textBackgroundColor = configManager.drawTextBackgroundColor
                }
                if let textCornerRadius = item["textCornerRadius"] as? CGFloat {
                    drawItem.textCornerRadius = textCornerRadius
                } else {
                    drawItem.textCornerRadius = configManager.drawTextCornerRadius
                }
                // Optional per-item font size for text annotations
                if let fontSize = item["fontSize"] as? CGFloat {
                    drawItem.textFontSize = fontSize
                } else {
                    drawItem.textFontSize = configManager.candleTextFontSize
                }

                klineView.drawContext.drawItemList.append(drawItem)
            }
            klineView.drawContext.setNeedsDisplay()
        }

        klineView.reloadConfigManager(configManager)
        shotView.shotColor = configManager.shotBackgroundColor
        if configManager.shouldFixDraw {
            configManager.shouldFixDraw = false
            klineView.drawContext.fixDrawItemList()
        }
        if (configManager.shouldClearDraw) {
            configManager.drawType = .none
            configManager.shouldClearDraw = false
            klineView.drawContext.clearDrawItemList()
        }
    }
    
    private func convertLocation(_ location: CGPoint) -> CGPoint {
        var reloadLocation = location
        reloadLocation.x = max(min(reloadLocation.x, bounds.size.width), 0)
        reloadLocation.y = max(min(reloadLocation.y, bounds.size.height), 0)
//        reloadLocation.x += klineView.contentOffset.x
        reloadLocation = klineView.valuePointFromViewPoint(reloadLocation)
        return reloadLocation
    }
    
    override func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
        let view = super.hitTest(point, with: event)
        guard view == klineView else {
            return view
        }

        // When no drawing UI is active, let HTKLineView handle touches so the chart scrolls normally.
        if configManager.shouldReloadDrawItemIndex == HTDrawState.none.rawValue {
            return view
        }

        // If we are actively creating a drawing (line / rect / etc.), always route touches to
        // the container so drawing gestures don't cause the chart itself to scroll.
        if configManager.drawType != .none {
            return self
        }

        // In "show" / manage-drawings mode (JS may pass drawType = -1 which maps to `.none` here),
        // we only intercept touches that actually hit an existing drawing. All other touches go to
        // HTKLineView so the user can still scroll the chart.
        if HTDrawItem.canResponseLocation(
            klineView.drawContext.drawItemList,
            convertLocation(point),
            klineView
        ) != nil {
            return self
        } else {
            return klineView
        }
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        touchesGesture(touches, .began)
    }
    
    override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
        touchesGesture(touches, .changed)
    }
    
    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        touchesGesture(touches, .ended)
    }
    
    override func touchesCancelled(_ touches: Set<UITouch>, with event: UIEvent?) {
        touchesEnded(touches, with: event)
    }
    
    func touchesGesture(_ touched: Set<UITouch>, _ state: UIGestureRecognizerState) {
        guard var location = touched.first?.location(in: self) else {
            shotView.shotPoint = nil
            return
        }
        var previousLocation = touched.first?.previousLocation(in: self) ?? location
        location = convertLocation(location)
        previousLocation = convertLocation(previousLocation)
        
        let translation = CGPoint.init(x: location.x - previousLocation.x, y: location.y - previousLocation.y)
        
        klineView.drawContext.touchesGesture(location, translation, state)
        shotView.shotPoint = state != .ended ? touched.first?.location(in: self) : nil
    }

    /// For a given X-value (timestamp), find the candle whose id is closest and
    /// return either the bottom or top of its real body in value-space
    /// (min(open, close) when useTop is false; max(open, close) when true).
    /// This mirrors HTDrawContext.bodyBottomValue(forX:)/bodyTopValue(forX:) and is used
    /// when reconstructing candleMarker drawings from JS that only provide `x`.
    private func candleMarkerBodyValue(forX value: CGFloat, useTop: Bool) -> CGFloat {
        guard !configManager.modelArray.isEmpty else {
            return value
        }
        var closest = configManager.modelArray[0]
        var minDiff = abs(closest.id - value)
        for model in configManager.modelArray {
            let diff = abs(model.id - value)
            if diff < minDiff {
                minDiff = diff
                closest = model
            }
        }
        let bodyLow = closest.low
        let bodyHigh = closest.high
        return useTop ? bodyHigh : bodyLow
    }

}

File: android/src/main/java/com/github/fujianlian/klinechart/container/HTKLineContainerView.java
package com.github.fujianlian.klinechart.container;


import android.view.MotionEvent;
import android.view.ViewGroup;
import android.widget.RelativeLayout;
import android.widget.ScrollView;
import com.facebook.react.bridge.*;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.events.RCTEventEmitter;
import com.github.fujianlian.klinechart.HTKLineConfigManager;
import com.github.fujianlian.klinechart.KLineChartView;
import com.github.fujianlian.klinechart.RNKLineView;
import com.github.fujianlian.klinechart.formatter.DateFormatter;
import java.util.List;
import java.util.Map;


public class HTKLineContainerView extends RelativeLayout {

    private ThemedReactContext reactContext;

    public HTKLineConfigManager configManager = new HTKLineConfigManager();

    public KLineChartView klineView;

    public HTShotView shotView;

    public HTKLineContainerView(ThemedReactContext context) {
        super(context);
        this.reactContext = context;
        klineView = new KLineChartView(getContext(), configManager);
        klineView.setGridColumns(5);
        klineView.setGridRows(3);
        klineView.setChildDraw(0);
        klineView.setDateTimeFormatter(new DateFormatter());
        klineView.configManager = configManager;
        // When scrolling to the left edge, request more historical candles from JS
        klineView.setRefreshListener(new KLineChartView.KChartRefreshListener() {
            @Override
            public void onLoadMoreBegin(KLineChartView chart) {
                int id = HTKLineContainerView.this.getId();
                reactContext.getJSModule(RCTEventEmitter.class).receiveEvent(
                        id,
                        RNKLineView.onEndReachedKey,
                        Arguments.createMap()
                );
                // Immediately end loading state so scrolling isn't locked
                chart.refreshComplete();
            }
        });
        addView(klineView, new LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT));
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        ViewGroup willShotView = (ViewGroup)getParent();
        if (shotView == null) {
            shotView = new HTShotView(getContext(), willShotView);
            shotView.setEnabled(false);
            shotView.dimension = 300;
            RelativeLayout.LayoutParams layoutParams = new RelativeLayout.LayoutParams(shotView.dimension, shotView.dimension);
            layoutParams.setMargins(50, 50, 0, 0);
            ((ViewGroup)willShotView.getParent().getParent()).addView(shotView, layoutParams);
        }
    }

    public void reloadConfigManager() {
        klineView.changeMainDrawType(klineView.configManager.primaryStatus);
        klineView.changeSecondDrawType(klineView.configManager.secondStatus);
        klineView.setMainDrawLine(klineView.configManager.isMinute);
        klineView.setPointWidth(klineView.configManager.itemWidth);
        klineView.setCandleWidth(klineView.configManager.candleWidth);

        if (klineView.configManager.fontFamily.length() > 0) {
            klineView.setTextFontFamily(klineView.configManager.fontFamily);
        }
        klineView.setTextColor(klineView.configManager.textColor);
        klineView.setTextSize(klineView.configManager.rightTextFontSize);
        klineView.setMTextSize(klineView.configManager.candleTextFontSize);
        klineView.setMTextColor(klineView.configManager.candleTextColor);
        klineView.reloadColor();
        Boolean isEnd = klineView.getScrollOffset() >= klineView.getMaxScrollX();
        klineView.notifyChanged();
        if (isEnd || klineView.configManager.shouldScrollToEnd) {
            klineView.setScrollX(klineView.getMaxScrollX());
        }

        final int id = this.getId();
        configManager.onDrawItemDidTouch = new Callback() {
            @Override
            public void invoke(Object... args) {
                HTDrawItem drawItem = (HTDrawItem) args[0];
                int drawItemIndex = (int) args[1];
                configManager.shouldReloadDrawItemIndex = drawItemIndex;

                WritableMap map = Arguments.createMap();
                if (drawItem != null) {
                    int drawColor = drawItem.drawColor;
                    int alpha = (drawColor >> 24) & 0xFF;
                    int red = (drawColor >> 16) & 0xFF;
                    int green = (drawColor >> 8) & 0xFF;
                    int blue = (drawColor) & 0xFF;
                    WritableArray colorList = Arguments.createArray();

                    colorList.pushDouble(red / 255.0);
                    colorList.pushDouble(green / 255.0);
                    colorList.pushDouble(blue / 255.0);
                    colorList.pushDouble(alpha / 255.0);

                    map.putArray("drawColor", colorList);
                    map.putDouble("drawLineHeight", drawItem.drawLineHeight);
                    map.putDouble("drawDashWidth", drawItem.drawDashWidth);
                    map.putDouble("drawDashSpace", drawItem.drawDashSpace);
                    map.putBoolean("drawIsLock", drawItem.drawIsLock);
                }
                // Expose the index of the touched drawing item to React Native.
                map.putInt("index", drawItemIndex);
                map.putInt("shouldReloadDrawItemIndex", drawItemIndex);
                reactContext.getJSModule(RCTEventEmitter.class).receiveEvent(
                        id,
                        RNKLineView.onDrawItemDidTouchKey,
                        map
                );
            }
        };
        configManager.onDrawItemComplete = new Callback() {
            @Override
            public void invoke(Object... args) {
                HTDrawItem drawItem = null;
                int drawItemIndex = -1;
                if (args != null && args.length >= 2) {
                    try {
                        drawItem = (HTDrawItem) args[0];
                        drawItemIndex = (int) args[1];
                    } catch (ClassCastException ignored) {
                    }
                }

                WritableMap map = Arguments.createMap();
                if (drawItem != null) {
                    map.putInt("index", drawItemIndex);
                    map.putInt("drawType", drawItem.drawType.rawValue());
                    map.putInt("drawColor", drawItem.drawColor);
                    map.putDouble("drawLineHeight", drawItem.drawLineHeight);
                    map.putDouble("drawDashWidth", drawItem.drawDashWidth);
                    map.putDouble("drawDashSpace", drawItem.drawDashSpace);
                    map.putBoolean("drawIsLock", drawItem.drawIsLock);
                    map.putString("text", drawItem.text);
                    map.putInt("textColor", drawItem.textColor);
                    map.putInt("textBackgroundColor", drawItem.textBackgroundColor);
                    map.putDouble("textCornerRadius", drawItem.textCornerRadius);
                    // Expose per-item text font size (falls back to candleTextFontSize when 0 on native).
                    map.putDouble("fontSize",
                            drawItem.textFontSize > 0 ? drawItem.textFontSize : configManager.candleTextFontSize);

                    WritableArray pointArray = Arguments.createArray();
                    for (HTPoint point : drawItem.pointList) {
                        WritableMap pointMap = Arguments.createMap();
                        pointMap.putDouble("x", point.x);
                        pointMap.putDouble("y", point.y);
                        pointArray.pushMap(pointMap);
                    }
                    map.putArray("pointList", pointArray);
                }
                reactContext.getJSModule(RCTEventEmitter.class).receiveEvent(
                        id,
                        RNKLineView.onDrawItemCompleteKey,
                        map
                );
            }
        };
        configManager.onDrawItemMove = new Callback() {
            @Override
            public void invoke(Object... args) {
                HTDrawItem drawItem = null;
                int drawItemIndex = -1;
                if (args != null && args.length >= 2) {
                    try {
                        drawItem = (HTDrawItem) args[0];
                        drawItemIndex = (int) args[1];
                    } catch (ClassCastException ignored) {
                    }
                }

                if (drawItem == null) {
                    return;
                }

                WritableMap map = Arguments.createMap();
                map.putInt("index", drawItemIndex);
                map.putInt("drawType", drawItem.drawType.rawValue());
                map.putString("text", drawItem.text);

                WritableArray pointArray = Arguments.createArray();
                for (HTPoint point : drawItem.pointList) {
                    WritableMap pointMap = Arguments.createMap();
                    pointMap.putDouble("x", point.x);
                    pointMap.putDouble("y", point.y);
                    pointArray.pushMap(pointMap);
                }
                map.putArray("pointList", pointArray);

                reactContext.getJSModule(RCTEventEmitter.class).receiveEvent(
                        id,
                        RNKLineView.onDrawItemMoveKey,
                        map
                );
            }
        };
        configManager.onDrawPointComplete = new Callback() {
            @Override
            public void invoke(Object... args) {
                HTDrawItem drawItem = (HTDrawItem) args[0];
                WritableMap map = Arguments.createMap();
                map.putInt("pointCount", drawItem.pointList.size());
                reactContext.getJSModule(RCTEventEmitter.class).receiveEvent(
                        id,
                        RNKLineView.onDrawPointCompleteKey,
                        map
                );
            }
        };

        int reloadIndex = configManager.shouldReloadDrawItemIndex;
        if (reloadIndex >= 0 && reloadIndex < klineView.drawContext.drawItemList.size()) {
            HTDrawItem drawItem = klineView.drawContext.drawItemList.get(reloadIndex);
            drawItem.drawColor = configManager.drawColor;
            drawItem.drawLineHeight = configManager.drawLineHeight;
            drawItem.drawDashWidth = configManager.drawDashWidth;
            drawItem.drawDashSpace = configManager.drawDashSpace;
            drawItem.drawIsLock = configManager.drawIsLock;
            if (configManager.drawShouldTrash) {
                configManager.shouldReloadDrawItemIndex = HTDrawState.showPencil;
                klineView.drawContext.drawItemList.remove(reloadIndex);
                configManager.drawShouldTrash = false;
            }
            klineView.drawContext.invalidate();
        }


        if (configManager.shouldFixDraw) {
            configManager.shouldFixDraw = false;
            klineView.drawContext.fixDrawItemList();
        }
        if (configManager.shouldClearDraw) {
            configManager.shouldReloadDrawItemIndex = HTDrawState.none;
            configManager.shouldClearDraw = false;
            klineView.drawContext.clearDrawItemList();
        }

        // If a serialized drawing list was provided from React Native,
        // rebuild the native drawItemList from it (pre-insert drawings).
        if (configManager.drawItemList != null) {
            klineView.drawContext.clearDrawItemList();
            for (Object itemObject : configManager.drawItemList) {
                if (!(itemObject instanceof Map)) {
                    continue;
                }
                Map itemMap = (Map) itemObject;
                Object pointListObject = itemMap.get("pointList");
                if (!(pointListObject instanceof List)) {
                    continue;
                }
                List pointList = (List) pointListObject;
                if (pointList.size() == 0) {
                    continue;
                }

                // Draw type
                int rawType = 0;
                Object drawTypeObject = itemMap.get("drawType");
                if (drawTypeObject instanceof Number) {
                    rawType = ((Number) drawTypeObject).intValue();
                }
                HTDrawType drawType = HTDrawType.drawTypeFromRawValue(rawType);

                // First point (required)
                Object firstPointObject = pointList.get(0);
                if (!(firstPointObject instanceof Map)) {
                    continue;
                }
                Map firstPointMap = (Map) firstPointObject;
                Object xObject = firstPointMap.get("x");
                Object yObject = firstPointMap.get("y");
                if (!(xObject instanceof Number) || !(yObject instanceof Number)) {
                    continue;
                }
                HTPoint startPoint = new HTPoint(
                        ((Number) xObject).floatValue(),
                        ((Number) yObject).floatValue()
                );

                HTDrawItem drawItem = new HTDrawItem(drawType, startPoint);

                // Remaining points (optional)
                for (int i = 1; i < pointList.size(); i++) {
                    Object pointObject = pointList.get(i);
                    if (!(pointObject instanceof Map)) {
                        continue;
                    }
                    Map pointMap = (Map) pointObject;
                    Object pxObject = pointMap.get("x");
                    Object pyObject = pointMap.get("y");
                    if (!(pxObject instanceof Number) || !(pyObject instanceof Number)) {
                        continue;
                    }
                    drawItem.pointList.add(new HTPoint(
                            ((Number) pxObject).floatValue(),
                            ((Number) pyObject).floatValue()
                    ));
                }

                // Style properties: fall back to global drawList config when not provided
                Object colorObject = itemMap.get("drawColor");
                if (colorObject instanceof Number) {
                    drawItem.drawColor = ((Number) colorObject).intValue();
                } else {
                    drawItem.drawColor = configManager.drawColor;
                }
                Object lineHeightObject = itemMap.get("drawLineHeight");
                if (lineHeightObject instanceof Number) {
                    drawItem.drawLineHeight = ((Number) lineHeightObject).floatValue();
                } else {
                    drawItem.drawLineHeight = configManager.drawLineHeight;
                }
                Object dashWidthObject = itemMap.get("drawDashWidth");
                if (dashWidthObject instanceof Number) {
                    drawItem.drawDashWidth = ((Number) dashWidthObject).floatValue();
                } else {
                    drawItem.drawDashWidth = configManager.drawDashWidth;
                }
                Object dashSpaceObject = itemMap.get("drawDashSpace");
                if (dashSpaceObject instanceof Number) {
                    drawItem.drawDashSpace = ((Number) dashSpaceObject).floatValue();
                } else {
                    drawItem.drawDashSpace = configManager.drawDashSpace;
                }
                Object isLockObject = itemMap.get("drawIsLock");
                if (isLockObject instanceof Boolean) {
                    drawItem.drawIsLock = (Boolean) isLockObject;
                } else {
                    drawItem.drawIsLock = configManager.drawIsLock;
                }

                Object textObject = itemMap.get("text");
                if (textObject instanceof String) {
                    drawItem.text = (String) textObject;
                }

                Object textColorObject = itemMap.get("textColor");
                if (textColorObject instanceof Number) {
                    drawItem.textColor = ((Number) textColorObject).intValue();
                } else {
                    drawItem.textColor = configManager.drawTextColor;
                }
                Object textBackgroundColorObject = itemMap.get("textBackgroundColor");
                if (textBackgroundColorObject instanceof Number) {
                    drawItem.textBackgroundColor = ((Number) textBackgroundColorObject).intValue();
                } else {
                    drawItem.textBackgroundColor = configManager.drawTextBackgroundColor;
                }
                Object textCornerRadiusObject = itemMap.get("textCornerRadius");
                if (textCornerRadiusObject instanceof Number) {
                    drawItem.textCornerRadius = ((Number) textCornerRadiusObject).floatValue();
                } else {
                    drawItem.textCornerRadius = configManager.drawTextCornerRadius;
                }
                // Optional per-item font size for text annotations
                Object fontSizeObject = itemMap.get("fontSize");
                if (fontSizeObject instanceof Number) {
                    drawItem.textFontSize = ((Number) fontSizeObject).floatValue();
                } else {
                    drawItem.textFontSize = configManager.candleTextFontSize;
                }

                klineView.drawContext.drawItemList.add(drawItem);
            }
            klineView.drawContext.invalidate();
        }

    }

    private HTPoint convertLocation(HTPoint location) {
        HTPoint reloadLocation = new HTPoint(location.x, location.y);
        reloadLocation.x = Math.max(0, Math.min(reloadLocation.x, getWidth()));
        reloadLocation.y = Math.max(0, Math.min(reloadLocation.y, getHeight()));
//        reloadLocation.x += klineView.getScrollOffset();
        reloadLocation = klineView.valuePointFromViewPoint(reloadLocation);
        return reloadLocation;
    }


    @Override
    public boolean onInterceptTouchEvent(MotionEvent event) {
        int action = event.getActionMasked();

        // When no drawing UI is active, always let KLineChartView handle touch events
        // so that normal scrolling and zooming work as expected.
        if (configManager.shouldReloadDrawItemIndex == HTDrawState.none) {
            return false;
        }

        // If we are actively creating a drawing (line / rect / etc.), intercept all events
        // so the chart itself doesn't scroll while the user is drawing.
        if (configManager.drawType != HTDrawType.none) {
            return true;
        }

        // In "show" / manage-drawings mode (JS may pass drawType = -1 which maps to none
        // on native), we want the user to be able to scroll the chart freely and only
        // intercept gestures that actually hit an existing drawing.
        HTPoint location = new HTPoint(event.getX(), event.getY());
        location = convertLocation(location);
        boolean hitExisting =
                HTDrawItem.canResponseLocation(klineView.drawContext.drawItemList, location, klineView) != null;

        switch (action) {
            case MotionEvent.ACTION_DOWN:
                // Start handling the gesture only if the user touched a drawing item.
                return hitExisting;
            case MotionEvent.ACTION_MOVE:
            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_CANCEL:
                // For move/up, keep the same routing decision as for ACTION_DOWN:
                // if the gesture started on a drawing, we keep intercepting; otherwise
                // let KLineChartView continue handling it for scrolling.
                return hitExisting;
            default:
                return false;
        }
    }

    private HTPoint lastLocation;

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        handlerDraw(event);
        handlerShot(event);
        return true;
    }

    private void handlerDraw(MotionEvent event) {
        HTPoint location = new HTPoint(event.getX(), event.getY());
        location = convertLocation(location);
        HTPoint previousLocation = lastLocation != null ? lastLocation : location;
        lastLocation = location;
        int state = event.getAction();
        Boolean isCancel = state == MotionEvent.ACTION_CANCEL;
        if (isCancel) {
            state = MotionEvent.ACTION_UP;
        }
        HTPoint translation = new HTPoint(
                location.x - previousLocation.x,
                location.y - previousLocation.y
        );
        if (event.getAction() == MotionEvent.ACTION_UP || event.getAction() == MotionEvent.ACTION_CANCEL) {
            lastLocation = null;
        }
        klineView.drawContext.touchesGesture(location, translation, state);
    }

    private void handlerShot(MotionEvent event) {
        if (event.getAction() == MotionEvent.ACTION_UP || event.getAction() == MotionEvent.ACTION_CANCEL) {
            shotView.setPoint(null);
            lastLocation = null;
        } else {
            shotView.setPoint(new HTPoint(event.getX(), event.getY()));
        }
    }

}
